include/nbl/builtin/hlsl/workgroup/fft.hlsl

```diff
@@ -1,13 +1,89 @@
+#include <nbl/builtin/hlsl/cpp_compat.hlsl>
+#include <nbl/builtin/hlsl/concepts.hlsl>
+#include <nbl/builtin/hlsl/fft/common.hlsl>
+
 #ifndef _NBL_BUILTIN_HLSL_WORKGROUP_FFT_INCLUDED_
 #define _NBL_BUILTIN_HLSL_WORKGROUP_FFT_INCLUDED_
 
+// ------------------------------- COMMON -----------------------------------------
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace workgroup
+{
+namespace fft
+{
+
+template<uint16_t _ElementsPerInvocationLog2, uint16_t _WorkgroupSizeLog2, typename _Scalar NBL_PRIMARY_REQUIRES(_ElementsPerInvocationLog2 > 0 && _WorkgroupSizeLog2 >= 5)
+struct ConstevalParameters
+{
+    using scalar_t = _Scalar;
+
+    NBL_CONSTEXPR_STATIC_INLINE uint16_t ElementsPerInvocationLog2 = _ElementsPerInvocationLog2;
+    NBL_CONSTEXPR_STATIC_INLINE uint16_t WorkgroupSizeLog2 = _WorkgroupSizeLog2;
+    NBL_CONSTEXPR_STATIC_INLINE uint32_t TotalSize = uint32_t(1) << (ElementsPerInvocationLog2 + WorkgroupSizeLog2);
+
+    NBL_CONSTEXPR_STATIC_INLINE uint16_t ElementsPerInvocation = uint16_t(1) << ElementsPerInvocationLog2;
+    NBL_CONSTEXPR_STATIC_INLINE uint16_t WorkgroupSize = uint16_t(1) << WorkgroupSizeLog2;
+
+    // Required size (in number of uint32_t elements) of the workgroup shared memory array needed for the FFT
+    NBL_CONSTEXPR_STATIC_INLINE uint32_t SharedMemoryDWORDs = (sizeof(complex_t<scalar_t>) / sizeof(uint32_t)) << WorkgroupSizeLog2;
+};
+
+}
+}
+}
+} 
+// ------------------------------- END COMMON ---------------------------------------------
+
+// -------------------------------- CPP ONLY ----------------------------------------------
+
+#ifndef __HLSL_VERSION
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace workgroup
+{
+namespace fft
+{
+
+struct OptimalFFTParameters
+{
+    uint16_t elementsPerInvocationLog2;
+    uint16_t workgroupSizeLog2;
+};
+
+inline OptimalFFTParameters optimalFFTParameters(const uint32_t maxWorkgroupSize, uint32_t inputArrayLength)
+{
```

Comment:
you need to return {0,0} (invalid value) if the workgroupsize is not at least 4 or the input array is less than 2

Reply:
Actually the inputArray needs to be at least twice the subgroupsize of the device, since we require full subgroups

Reply:
Not sure how to assert that from cpp side unless I make the methods divergent between host and device

Reply:
> Actually the inputArray needs to be at least twice the subgroupsize of the device, since we require full subgroups

The thing is you can't run a workgroup smaller than a subgroup, so whenever FFT input array is less than that, the input signal should be padded with periodically repeated copies of itself (it makes the FFT output contain 0s in the "padding" frequencies)

Reply:
> Not sure how to assert that from cpp side unless I make the methods divergent between host and device

take a ~~minWorkgroupSize~~ minSubgroupSize and a maxWorkgroupSize.

The min should be user provided.